File: api/backend/python/xl-client/dial_xl/reader.py
import bisect
from typing import Any, Callable


class _Reader:
    """A helper class to read parsed DSL ranges."""

    dsl: str
    all_positions: list[int]
    position: int
    entity: Any

    def __init__(
        self, dsl: str, all_positions: list[int], entity: Any, position: int
    ):
        self.dsl = dsl
        self.all_positions = all_positions
        self.entity = entity
        self.position = position

    def next_unparsed(self, to: Callable[[Any], int]) -> str:
        next_position = to(self.entity)
        unparsed_end = next(
            (
                i + 1
                for i in reversed(range(self.position, next_position))
                if not self.dsl[i].isspace()
            ),
            self.position,
        )
        result = self.dsl[self.position : unparsed_end]
        self.position = unparsed_end

        return result

    def next(self, to: int | Callable[[Any], int]) -> str:
        next_position = to if isinstance(to, int) else to(self.entity)
        result = self.dsl[self.position : next_position]
        self.position = next_position

        return result

    def before_next(self) -> str:
        index = bisect.bisect_left(self.all_positions, self.position)
        next_position = (
            self.all_positions[index]
            if index < len(self.all_positions)
            else self.position
        )
        result = self.dsl[self.position : next_position]
        self.position = next_position

        return result

    def till_linebreak(self, stop_on_text: bool = False) -> str:
        stop = self._get_stop(stop_on_text)
        index = self.dsl.find("\n", self.position, stop) + 1
        next_position = index if index > 0 else stop
        result = self.dsl[self.position : next_position]
        self.position = next_position

        return result

    def with_entity(self, entity: Any) -> "_Reader":
        return _Reader(self.dsl, self.all_positions, entity, self.position)

    def _get_stop(self, stop_on_text: bool) -> int:
        if stop_on_text:
            return next(
                (
                    i
                    for i in range(self.position, len(self.dsl))
                    if not self.dsl[i].isspace()
                ),
                len(self.dsl),
            )
        else:
            index = bisect.bisect_right(self.all_positions, self.position)
            return (
                self.all_positions[index]
                if index < len(self.all_positions)
                else self.position
            )

File: frontend/libs/parser/src/lib/EditDslApi/utils/Reader.ts
import { lineBreak } from '../../parser';

/**
 * A helper class to read DSL ranges.
 */
export class Reader {
  public dsl: string;
  public allPositions: number[];
  public entity: any;
  public position: number;

  constructor(
    dsl: string,
    allPositions: number[],
    entity: any,
    position: number
  ) {
    this.dsl = dsl;
    this.allPositions = allPositions;
    this.entity = entity;
    this.position = position;
  }

  /**
   * Reads text up to the position returned by `to(entity)`,
   * trimming trailing whitespace from the end.
   */
  public nextUnparsed(to: (entity: any) => number): string {
    const nextPosition = to(this.entity);
    // Find the last non-whitespace character before nextPosition
    let unparsedEnd = this.position;
    for (let i = nextPosition - 1; i >= this.position; i--) {
      if (!/\s/.test(this.dsl[i])) {
        unparsedEnd = i + 1;
        break;
      }
    }

    const result = this.dsl.substring(this.position, unparsedEnd);
    this.position = unparsedEnd;

    return result;
  }

  /**
   * Reads text from the current position up to `to`, which can be either
   * a number or a function returning a number.
   */
  public next(to: number | ((entity: any) => number)): string {
    const nextPosition = typeof to === 'number' ? to : to(this.entity);
    const result = this.dsl.substring(this.position, nextPosition);
    this.position = nextPosition;

    return result;
  }

  /**
   * Reads text from the current position to the next relevant "break" in `allPositions`
   */
  public beforeNext(): string {
    const index = bisectLeft(this.allPositions, this.position);
    const nextPosition =
      index < this.allPositions.length
        ? this.allPositions[index]
        : this.position;

    const result = this.dsl.substring(this.position, nextPosition);
    this.position = nextPosition;

    return result;
  }

  /**
   * Reads text from the current position until the next line break,
   * or until the next "stop" position in `allPositions`, depending
   * on `stop_on_text`.
   */
  public tillLinebreak(stopOnText = false): string {
    const stop = this.getStop(stopOnText);
    const index = this.dsl.indexOf(lineBreak, this.position) + 1; // +1 to include the newline
    const endOfLine = index > 0 && index <= stop ? index : stop;

    const result = this.dsl.substring(this.position, endOfLine);
    this.position = endOfLine;

    return result;
  }

  /**
   * Returns a new Reader instance with the same DSL, positions, and
   * position index, but a different `entity`.
   */
  public withEntity(entity: any): Reader {
    return new Reader(this.dsl, this.allPositions, entity, this.position);
  }

  /**
   * Internal helper to calculate where reading should stop
   * for `till_linebreak`.
   */
  private getStop(stopOnText: boolean): number {
    if (stopOnText) {
      // Find the first non-whitespace character from current position
      // or use the end of the DSL if none is found
      for (let i = this.position; i < this.dsl.length; i++) {
        if (!/\s/.test(this.dsl[i])) {
          return i;
        }
      }

      return this.dsl.length;
    } else {
      const index = bisectRight(this.allPositions, this.position);

      return index < this.allPositions.length
        ? this.allPositions[index]
        : this.position;
    }
  }
}

/**
 * Returns the index at which `x` should be inserted into `arr`
 * (which is assumed to be sorted) in order to maintain sorted order.
 * If `x` is already present in `arr`, the insertion point will be
 * to the left of any existing entries.
 *
 * @param arr - A sorted array of numbers.
 * @param x - The value to insert.
 * @returns The insertion index, an integer in the range [0, arr.length].
 */
function bisectLeft(arr: number[], x: number): number {
  let low = 0;
  let high = arr.length;
  while (low < high) {
    const mid = (low + high) >>> 1;
    if (arr[mid] < x) {
      low = mid + 1;
    } else {
      high = mid;
    }
  }

  return low;
}

/**
 * Returns the index at which `x` should be inserted into `arr`
 * (which is assumed to be sorted) in order to maintain sorted order.
 * If `x` is already present in `arr`, the insertion point will be
 * to the right of any existing entries.
 *
 * @param arr - A sorted array of numbers.
 * @param x - The value to insert.
 * @returns The insertion index, an integer in the range [0, arr.length].
 */
function bisectRight(arr: number[], x: number): number {
  let low = 0;
  let high = arr.length;
  while (low < high) {
    const mid = (low + high) >>> 1;
    if (arr[mid] <= x) {
      low = mid + 1;
    } else {
      high = mid;
    }
  }

  return low;
}
